hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/CyclicWriteBehindQueue.java

```diff
@@ -0,0 +1,222 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl.mapstore.writebehind;
+
+import com.hazelcast.nio.serialization.Data;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static com.hazelcast.util.ValidationUtil.checkNotNull;
+
+/**
+ * Write behind queue impl. backed by a {@link java.util.Deque}.
+ * Used when non-write-coalescing mode is on.
+ * Intended to use when we need to store all changes on a key.
+ */
+class CyclicWriteBehindQueue implements WriteBehindQueue<DelayedEntry> {
+
+    private final Deque<DelayedEntry> deque;
+
+    /**
+     * Index to fast check existence of a key.
+     * Maps: key --> number of keys.
+     * <p/>
+     * Used to determine whether a key is loadable from store. Because there is a possibility that a key
+     * is in {@link WriteBehindQueue} but it is not in {@link com.hazelcast.core.IMap} due to the eviction.
+     * At that point if one tries to get that evicted key, {@link com.hazelcast.core.MapLoader} will
+     * try to load it from store and that may cause data inconsistencies.
+     *
+     * @see WriteBehindStore#loadable(com.hazelcast.nio.serialization.Data)
+     */
+    private final Map<Data, Integer> index;
+
+    public CyclicWriteBehindQueue() {
+        this.deque = new ArrayDeque<DelayedEntry>();
+        this.index = new HashMap<Data, Integer>();
+    }
+
+    /**
+     * Inserts to the end of this queue.
+     *
+     * @param entry item to be offered
+     */
+    @Override
+    public void addEnd(DelayedEntry entry) {
+        deque.addLast(entry);
+        addCountIndex(entry);
+    }
+
+    /**
+     * Add this collection to the front of the queue.
+     *
+     * @param collection collection to be added in front of this queue.
+     */
+    @Override
+    public void addFront(Collection<DelayedEntry> collection) {
+        for (DelayedEntry entry : collection) {
+            deque.addFirst(entry);
+        }
+        addCountIndex(collection);
+    }
+
+    /**
+     * Removes all elements matching with the elements in the supplied collection.
+     *
+     * @param entry element to be removed.
+     * @return <code>true</code> if removed successfully, <code>false</code> otherwise
+     */
+    @Override
+    public boolean removeFirstOccurrence(DelayedEntry entry) {
```

Comment:
I didn't get how this method removes first occurrence, looks like it just polls first.
maybe I didn't get the logic behind this class